### PR TITLE
[AutoDiff] Remove differentiation mode.

### PR DESCRIPTION
**Pull request by dan-zheng**
Differentiation mode (`forward` or `reverse`) is not necessary.
We no longer *differentiate* the two (pun by @rxwei).
Addresses [SR-9487](https://bugs.swift.org/browse/SR-9487).

Note: sometimes the attribute is spelled `@differentiable()`, the parentheses are currently required. We probably want to support `@differentiable` in favor of that.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
Sorry, please hold off on reviewing. I need to fix some comma-parsing logic.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Review by rxwei — APPROVED** [review]
Great. Thanks!

**Comment by dan-zheng** [comment]
Note: unfortunately, removing differentiation mode somewhat complicates parsing, specifically diagnostics.

Previously, mode was the first required component in the `@differentiable` attribute.
Parsing worked as follows:

```swift
// Invalid attribute example.
@differentiable(reverse, 8)

// 1. Parse mode up until comma after mode:
                ~~~~~~~^
//    Note that this first comma is guaranteed to appear in any valid
//    `@differentiable` declaration.

// 2. Attempt to parse 'wrt:' clause, which has a leading comma:
//        ',' 'wrt' ':' ...
//    Then, attempt to parse four function specifiers (primal, adjoint, jvp, vjp).
//    Each specifier has a leading comma:
//        ',' 'primal|adjoint|jvp|vjp' ':' decl-name

// 3. At this point, all valid commas should have been parsed.
//    Check token. If it's still a comma, then the next token ('8' in this case) is bad.
//    Emit diagnostic: "expected a configuration, e.g. 'wrt:', 'primal:', ..."
```

Now that mode has been removed, there's no guaranteed "first comma".

Thus, the parsing logic for function specifiers needs to change.
Previously, all function specifiers were guaranteed to have a leading comma.
Now, they only sometimes have a leading comma. (hence the gnarly `tokIsIdentifier` lambda in `parseDifferentiableAttributeArguments`)

Also, the condition for emitting a diagnostic needs to change.
The new condition is (after attempting to parse 'wrt:' and function specifiers):
- If token is a comma, or if parser has not advanced after parsing the opening '(',
   then the next token is bad.

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -896,43 +892,43 @@ bool Parser::parseDifferentiableAttributeArguments(
     return true;
   };
 
+  // Parse trailing comma, if it exists, and check for errors.
+  auto consumeIfTrailingComma = [&]() -> bool {
+    if (consumeIf(tok::comma)) {
```
Comment (rxwei) [inline lib/Parse/ParseDecl.cpp]: Negate the condition and early return.

File: lib/SILGen/SILGen.cpp [hunk 2]
```diff
@@ -732,46 +732,37 @@ void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
...
+      jvpName = getFunction(SILDeclRef(jvpFn), ForDefinition)->getName();
+    if (auto *vjpFn = diffAttr->getVJPFunction())
+      vjpName = getFunction(SILDeclRef(vjpFn), ForDefinition)->getName();
+    // Get lowered argument indices.
+    auto paramIndices = diffAttr->getCheckedParameterIndices()->getLowered(
+        AFD->getInterfaceType()->castTo<AnyFunctionType>());
+    SILAutoDiffIndices indices(/*source*/ 0, paramIndices);
+    silOriginalFn->addDifferentiableAttr(
+        SILDifferentiableAttr::create(
+          M, indices, primName, adjName,
```
Comment (rxwei) [inline lib/SILGen/SILGen.cpp]:
```suggestion
            M, indices, primName, adjName,
```

File: lib/SILGen/SILGen.cpp [hunk 3]
```diff
@@ -732,46 +732,37 @@ void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
...
+          /*primitive*/ hasPrimitiveAdjoint, jvpName, vjpName));
```
Comment (rxwei) [inline lib/SILGen/SILGen.cpp]:
```suggestion
            /*primitive*/ hasPrimitiveAdjoint, jvpName, vjpName));
```

**Review by rxwei — APPROVED** [review]

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/Parse/ParseDecl.cpp]: Thanks! Done in https://github.com/apple/swift/pull/21544/commits/adff0ab2125678960de1c27ba1d6eeede181b545

File: lib/SILGen/SILGen.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILGen/SILGen.cpp]: Done in https://github.com/apple/swift/pull/21544/commits/adff0ab2125678960de1c27ba1d6eeede181b545

File: lib/SILGen/SILGen.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/SILGen/SILGen.cpp]: Done in https://github.com/apple/swift/pull/21544/commits/adff0ab2125678960de1c27ba1d6eeede181b545

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**